src/NHibernate/Util/AsyncReaderWriterLock.cs

```diff
@@ -0,0 +1,229 @@
+
+		public Releaser WriteLock()
+		{
+			if (!CanEnterWriteLock(out var waitForReadLocks))
+			{
+				_writeLockSemaphore.Wait();
+				lock (_writeLockSemaphore)
+				{
+					_writersWaiting--;
+				}
+			}
+
+			if (waitForReadLocks)
+			{
+				_readLockSemaphore.Wait();
+			}
+
+			return _writerReleaser;
+		}
+
+		public async Task<Releaser> WriteLockAsync()
+		{
+			if (!CanEnterWriteLock(out var waitForReadLocks))
+			{
+				await _writeLockSemaphore.WaitAsync().ConfigureAwait(false);
+				lock (_writeLockSemaphore)
+				{
+					_writersWaiting--;
+				}
+			}
+
+			if (waitForReadLocks)
+			{
+				await _readLockSemaphore.WaitAsync().ConfigureAwait(false);
+			}
+
+			return _writerReleaser;
+		}
+
+		public Releaser ReadLock()
+		{
+			if (CanEnterReadLock())
+			{
+				return _readerReleaser;
+			}
+
+			_waitingReadLockSemaphore.Wait();
+			ReleaseWaitingReader();
+
+			return _readerReleaser;
+		}
+
+		public Task<Releaser> ReadLockAsync()
+		{
+			return CanEnterReadLock() ? _readerReleaserTask : ReadLockInternalAsync();
+
+			async Task<Releaser> ReadLockInternalAsync()
+			{
+				await _waitingReadLockSemaphore.WaitAsync().ConfigureAwait(false);
+				ReleaseWaitingReader();
+
+				return _readerReleaser;
+			}
+		}
+
+		private bool CanEnterWriteLock(out bool waitForReadLocks)
+		{
+			waitForReadLocks = false;
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount > 0 && _writeLockSemaphore.Wait(0))
+				{
+					waitForReadLocks = _currentReaders > 0;
+					return true;
+				}
+
+				_writersWaiting++;
+			}
+
+			return false;
+		}
+
+		private void ExitWriteLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount == 1)
+				{
+					throw new InvalidOperationException();
+				}
+
+				// Writers have the highest priority even if they came last
+				if (_writersWaiting > 0 || _waitingReadLockSemaphore == null)
+				{
+					_writeLockSemaphore.Release();
+					return;
+				}
+
+				if (_readersWaiting > 0)
+				{
+					_currentReaders += _readersWaiting;
+					_waitingReadLockSemaphore.Release(_readersWaiting);
+				}
+
+				_writeLockSemaphore.Release();
+			}
+		}
+
+		private bool CanEnterReadLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writersWaiting == 0 && _writeLockSemaphore.CurrentCount > 0)
+				{
+					_currentReaders++;
+
+					return true;
+				}
+
+				if (_waitingReadLockSemaphore == null)
+				{
+					_waitingReadLockSemaphore = new SemaphoreSlim(0);
+				}
+
+				_readersWaiting++;
+
+				return false;
+			}
+		}
+
+		private void ExitReadLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_currentReaders == 0)
+				{
+					throw new InvalidOperationException();
+				}
+
+				_currentReaders--;
+				if (_currentReaders == 0 && _writeLockSemaphore.CurrentCount == 0)
+				{
+					_readLockSemaphore.Release();
+				}
+			}
+		}
+
+		private void ReleaseWaitingReader()
+		{
+			lock (_writeLockSemaphore)
```

Comment:
This creates a synchronization point between all released readers. If think it can be avoided, but I am not sure it is worth it.

In the `ExitWriteLock`, we could store the `_waitingReadLockSemaphore` into a `_waitingDisposalSemaphore` member, release the readers, set `_readersWaiting` directly to 0 and `_waitingReadLockSemaphore` to `null`. `ReleaseWaitingReader` will be dropped.

In `WriteLock` and in `WriteLockAsync`, after the `if (waitForReadLocks) { }`, a non-null `_waitingDisposalSemaphore` would be the disposed and nullified. At that point, we are certain that all readers which were waiting in this semaphore have finished their work, and so it is safe to dispose it.
This is in case we want to do it at the soonest possible point without having to lock for all released readers. Otherwise this disposal could be done in `ExitWriteLock` before storing the current `_waitingReadLockSemaphore` into `_waitingDisposalSemaphore`.

Reply:
> If think it can be avoided, but I am not sure it is worth it.

It is worth it as it prevents from reusing the old waiting reader lock.